File: cpp/Dynamic Programming/L_largestDivisbleSubset.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> largestDivisibleSubset(vector<int> &nums)
{
    sort(nums.begin(), nums.end());
    int n = nums.size();
    vector<int> groupSize(n, 1), prevElement(n, -1);
    int maxIndex = 0;

    for (int i = 0; i < n; ++i)
    {
        for (int j = 0; j < i; ++j)
        {
            if (nums[i] % nums[j] == 0 && groupSize[i] < groupSize[j] + 1)
            {
                groupSize[i] = groupSize[j] + 1;
                prevElement[i] = j;
            }
        }
        if (groupSize[i] > groupSize[maxIndex])
        {
            maxIndex = i;
        }
    }

    vector<int> result;
    for (int i = maxIndex; i != -1; i = prevElement[i])
    {
        result.insert(result.begin(), nums[i]);
    }
    return result;
}

int n;
vector<int> ans;
void solve(vector<int> &nums, int i, int j, vector<int> cur)
{
    if (j >= n)
    {
        if (cur.size() > ans.size())
            ans = cur;
        return;
    }
    for (int k = j; k < n; k++)
    {
        if (nums[k] % nums[i] == 0)
        {
            cur.push_back(nums[k]);
            solve(nums, k, k + 1, cur);
            cur.pop_back();
        }
    }
    if (cur.size() > ans.size())
        ans = cur;
}
vector<int> largestDivisibleSubset(vector<int> &nums)
{
    n = nums.size();
    int mx = 0;
    sort(nums.begin(), nums.end());
    vector<int> cur;
    for (int i = 0; i < n; i++)
    {
        solve(nums, i, i + 1, {nums[i]});
    }
    return ans;
}

File: Python/Dynamic Programming/L_largestDivisbleSubset.py
from typing import List


def largestDivisibleSubset(nums: List[int]) -> List[int]:
    nums.sort()
    n = len(nums)
    groupSize = [1] * n
    prevElement = [-1] * n
    maxIndex = 0

    for i in range(n):
        for j in range(i):
            if nums[i] % nums[j] == 0:
                if groupSize[i] < 1 + groupSize[j]:
                    groupSize[i] = 1 + groupSize[j]
                    prevElement[i] = j
        if groupSize[i] > groupSize[maxIndex]:
            maxIndex = i

    result = []
    while maxIndex != -1:
        result.insert(0, nums[maxIndex])
        maxIndex = prevElement[maxIndex]

    return result

def largestDivisibleSubset1(nums: List[int]) -> List[int]:
    nums.sort()
    n = len(nums)
    dp = [1] * n
    max_size, max_index = 1, 0

    for i in range(1, n):
        for j in range(i):
            if nums[i] % nums[j] == 0:
                dp[i] = max(dp[i], dp[j] + 1)
                if dp[i] > max_size:
                    max_size = dp[i]
                    max_index = i

    result = []
    num = nums[max_index]
    for i in range(max_index, -1, -1):
        if num % nums[i] == 0 and dp[i] == max_size:
            result.append(nums[i])
            num = nums[i]
            max_size -= 1

    return result


